exe/main.cc

```diff
@@ -40,90 +53,118 @@ int startAgent(std::vector<std::string> arguments) {
         }
     } catch (const std::exception& e) {
         LOG_ERROR("Failed to daemonize: %1%", e.what());
-        return 1;
+        return EXIT_FAILURE;
     } catch (...) {
         LOG_ERROR("Failed to daemonize");
-        return 1;
-    }
-
-    if (!Configuration::Instance().isInitialized()) {
-        // Start a thread that just busy waits to facilitate
-        // acceptance testing
-        PCPClient::Util::thread idle_thread {
-            []() {
-                LOG_WARNING("pxp-agent started unconfigured. No connection can "
-                            "be established");
-                for (;;) {
-                    PCPClient::Util::this_thread::sleep_for(PCPClient::Util::chrono::milliseconds(5000));
-                }
-            } };
-        idle_thread.join();
-        return 0;
+        return EXIT_FAILURE;
     }
 
-    bool success { false };
-    const auto& agent_configuration =
-        Configuration::Instance().getAgentConfiguration();
-
-    try {
-        Agent agent { agent_configuration };
-        agent.start();
-        success = true;
-    } catch (const Agent::Error& e) {
-        LOG_ERROR("Fatal error: %1%", e.what());
-    } catch (const std::exception& e) {
-        LOG_ERROR("Unexpected error: %1%", e.what());
-    } catch (...) {
-        LOG_ERROR("Unexpected error");
+    bool success { true };
+    if (!Configuration::Instance().valid()) {
+        // pxp-agent will execute in uncofigured mode
+        loopIdly();
+    } else {
+        try {
+            Agent agent { Configuration::Instance().getAgentConfiguration() };
+            agent.start();
+        } catch (const Agent::WebSocketConfigurationError& e) {
+            LOG_ERROR("WebSocket configuration error (%1%) - pxp-agent will "
+                      "continue executing, but will not attempt to connect to "
+                      "the PCP broker again", e.what());
+            loopIdly();
+        } catch (const Agent::FatalError& e) {
+            success = false;
+            LOG_ERROR("Fatal error: %1%", e.what());
+        } catch (const std::exception& e) {
+            success = false;
+            LOG_ERROR("Unexpected error: %1%", e.what());
+        } catch (...) {
+            success = false;
+            LOG_ERROR("Unexpected error");
+        }
     }
 
 #ifdef _WIN32
     Util::daemon_cleanup();
 #endif
-    return (success ? 0 : 1);
+    return (success ? EXIT_SUCCESS : EXIT_FAILURE);
 }
 
 int main(int argc, char *argv[]) {
+    // Initialize
+
     // Fix args on Windows to be UTF-8
     boost::nowide::args arg_utf8(argc, argv);
 
-    Configuration::Instance().setStartFunction(startAgent);
+    Configuration::Instance().initialize(startAgent);
+
+    // Parse options
 
     HW::ParseResult parse_result { HW::ParseResult::OK };
+    std::string err_msg {};
 
     try {
-        parse_result = Configuration::Instance().initialize(argc, argv);
+        parse_result = Configuration::Instance().parseOptions(argc, argv);
```

Comment:
Separating the 1) parse step from 2) setup logging and 3) validate.
That helps understanding why the logging can be set up (useful for PCP-134).